### ISSUE TITLE
Cookies set in thrown responses are ignored when the CatchBoundary is in a parent route

### ISSUE THREAD

**Issue by eranhirsch** [remix-run/remix#3157]
### What version of Remix are you using?

1.4.3

### Steps to Reproduce

1. Create a new remix app using `npx create-remix` and use the `Just the basics` and `Remix App Server` settings
2. Create the **parent layout** `~/app/routes/parent.tsx`:
```js
export function CatchBoundary() {
  return <main>I'm a catch</main>;
}

export default function Parent() {
  return (
    <main>
      <header>Parent</header>
      <Outlet />
    </main>
  );
}
```
3. Create the **child** `~/app/routes/parent/child.tsx`:
```js
export const loader: LoaderFunction = ({ request }) => {
  if (new URL(request.url).searchParams.has("throw")) {
    throw json(null, { headers: { "Set-Cookie": "thrownCookie=DoesntWork" } });
  }
  return json(null, { headers: { "Set-Cookie": "regularFlowCookie=ThisWorks" } });
};

export default function Child() {
  return <section>Child</section>;
}
```
4. Run the server `npm run dev`
5. `curl --head "localhost:3000/parent/child"`
6. `curl --head "localhost:3000/parent/child?throw"`

### Expected Behavior

The **Set-Cookie** header should be shown in both cases

### Actual Behavior

When the response is thrown in a child of a CatchBoundary, **Set-Cookie** headers are not copied

**Comment by eranhirsch** [comment]
This happens because of 
https://github.com/remix-run/remix/blob/main/packages/remix-server-runtime/server.ts#L404
which filters out the modules up the CatchBoundary.

Because that filtered prefix of routes is the one used to generate the headers:
https://github.com/remix-run/remix/blob/main/packages/remix-server-runtime/server.ts#L434-L439

The problem is that, at this stage of the response logic, we don't track that a loader response was thrown or if it was returned in a happy flow (except logging the catch itself and the response *data* and *status*). So while it makes sense to skip the data (and headers) of child loaders that ran successfully, as they would be irrelevant because of the CatchBoundary at their parent, the thrown response should be treated as if it is "rendered" (it is what the CatchBoundary "catches") so it's headers should be handled too.

**Comment by brophdawg11** [comment]
Fixed by #6475 and will be available in the next release (likely 1.17.0)

**Comment by eranhirsch** [comment]
wow, it's been a year! thanks for circling back to the task to update! Waiting for the next release to remove my patched solution
